### ISSUE TITLE
withAuthenticationRequired causing double component mount

### ISSUE THREAD

**Issue by pelkiegr** [auth0/auth0-react#382]
### Describe the problem

> The `withAuthenticationRequired` HOC causes a double render when changing context above the `Route` tree (i.e. theme).  This causes the route to re-initialize and lose all of it's current state.

### What was the expected behavior?

> Components to only mount once.

### Reproduction

> Detail the steps taken to reproduce this error, and whether this issue can be reproduced consistently or if it is intermittent.
> **Note**: If clear, reproducible steps or the smallest sample app demonstrating misbehavior cannot be provided, we may not be able to follow up on this bug report.

- Clone `https://github.com/auth0-developer-hub/spa_react_typescript_hello-world_react-router-6.git --branch basic-authentication` as mentioned in [this blog post ](https://auth0.com/developers/hub/code-samples/spa/react-typescript/basic-authentication-react-router-6) 
(same behavior in the `basic-authentication-bug` branch)
- Create an `ApplicationContext`
```
import React, {createContext, Dispatch, SetStateAction, useMemo, useState} from "react";

interface ContextProps {
  toggle: boolean;
  setToggle: Dispatch<SetStateAction<boolean>>;
}

const defaultState = {
  toggle: false,
  setToggle: () => null,
};

export const AppContext = createContext<ContextProps>(defaultState);

const AppContextProvider: React.FC = ({ children }) => {

  const [toggle, setToggle] = useState(true);

  return <AppContext.Provider value={{toggle, setToggle}}>{children}</AppContext.Provider>;
};

export default AppContextProvider;
```
- Place `ApplicationContext` in the `index` around the app
- Add a button to the nav bar to consume the setter from the context
```
      <button onClick={() => setToggle((prev) => !prev)}>Toggle</button>
```
- Add this use effect in the profile route to track mounts/unmounts
```
  useEffect(() => {
    console.log('mount profile');
    return () => console.log('unmount profile');
  }, [])
```
- console log `toggle` from `App.tsx` to trigger a re-render
- Toggle the button and see that the `Profile` component is unmounted and remounted.  (removing the `withAuthenticationRequired` lines from `ProtectedRoute` and returning component directly gives the expected behavior)

### Environment

> Please provide the following:

    "@auth0/auth0-react": "^1.9.0",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-router-dom": "^6.2.1",

### Additional Information
 I have tried downgrading to React Router 5 as well as most of the solutions listed in this related issue but see the same behavior.
- https://github.com/auth0/auth0-react/issues/335

**Comment by adamjmcgrath** [comment]
Hi @pelkiegr - thanks for raising this.

Am not able to reproduce the issue you describe using the `basic-authentication-bug` branch

I've created an app following your steps here https://github.com/adamjmcgrath/spa_react_typescript_hello-world_react-router-6/tree/basic-authentication-bug

The profile page only mounts once and and doesn't update when I update the "toggle" context

<img width="818" alt="image" src="https://user-images.githubusercontent.com/1299658/179966957-7d30c1ca-e5f0-4846-abf1-fc9811df5ee6.png">

**Comment by pelkiegr** [comment]
@adamjmcgrath Sorry, I think I missed one step (I should have just pushed my changes somewhere).

Anyway, in the branch you updated, if you add
`const { toggle } = useContext(AppContext);`
to
`App`, i.e.
```
export const App: React.FC = () => {
  const { isLoading } = useAuth0();
  const { toggle } = useContext(AppContext);
```

you can reproduce the behavior I'm talking about.

**Comment by adamjmcgrath** [comment]
@pelkiegr - thanks for sharing that

Causing the app to re-render will cause the Profile page to remount.

`withAuthenticationRequired` is a Higher Order Component, it's a function that returns a component - a new component each time it's called. 

React’s diffing algorithm uses component identity to determine whether it should update the existing subtree or throw it away and mount a new one. New component means new identity.

So using a HOC inside a render will cause a remount each time the parent component is rendered. See https://reactjs.org/docs/higher-order-components.html#dont-use-hocs-inside-the-render-method

The same would apply if you used any HOC (there's nothing special about `withAuthenticationRequired`) in that way, eg https://github.com/adamjmcgrath/spa_react_typescript_hello-world_react-router-6/compare/basic-authentication-bug...adamjmcgrath:spa_react_typescript_hello-world_react-router-6:any-hoc

Generally, you shouldn't use HOCs inside renders (unless you know the component wont be re-rendered)
eg.

```js
const HOC = (Component) => ({ ...props }) => <Component {...props}/>

// NOT OK
const App = () {
  const { toggle } = useContext(SomeContext);
  const WrappedProfile = HOC(Profile);
  return <WrappedProfile />;
}

// OK
const WrappedProfile = HOC(Profile);
const App = () {
  const { toggle } = useContext(SomeContext);
  return <WrappedProfile />;
}
```

In short, you need to move `withAuthenticationRequired` out of the render of the `App`, if you want to re-render the `App`. eg 

```js
const ProtectedProfile = withAuthenticationRequired(Profile);

export const App: React.FC = () => {

  // ...

  return (
    <BrowserRouter>
       // ...
           <Route
              path="/profile"
              element={<ProtectedProfile />}
            />
        // ...    
    </BrowserRouter>
  );
};

```

**Comment by pelkiegr** [comment]
@adamjmcgrath thank you - that makes sense.  I completely missed that.  I *think* another option would be to do something like
```
export default React.memo(ProtectedRoute);
```
inside of the protected route in the example.  I guess that comes with some additional overhead vs the approach you mentioned.

I wonder either way if it's worth updating the Auth0 example or at least adding a comment in the protected route file so others don't (naively) run into this issue in the future?

**Comment by adamjmcgrath** [comment]
> `export default React.memo(ProtectedRoute);`

Yep 👍 

> I wonder either way if it's worth updating the Auth0 example 

Yep, good call - have updated it https://github.com/auth0/auth0-react/pull/385
